File: AdventOfCode/2022/day13.js
const { part_one, part_two, sum, read } = require("../../common/aoc");

const messages_file = read("example");

const ordered_indices = messages_file
    .split("\n\n")
    .map((b) => b.split("\n").map((js) => JSON.parse(js)))
    .map(([a, b], i) => (comparePackets(a, b) < 0 ? i + 1 : 0));

part_one(sum(ordered_indices));

const orderedMessages = [
    ...messages_file
        .split("\n")
        .filter((s) => s.length > 0) // remove empty lines
        .map((js) => JSON.parse(js)),
    [[2]],
    [[6]],
]
    .sort(comparePackets)
    .map((s) => JSON.stringify(s));

part_two(
    (orderedMessages.findIndex((s) => s === "[[2]]") + 1) *
        (orderedMessages.findIndex((s) => s === "[[6]]") + 1)
);

function comparePackets(l, r) {
    if (Array.isArray(l) && Array.isArray(r)) {
        if (l.length === 0 && r.length > 0) return -1;
        else if (l.length === 0 && r.length === 0) return 0;
        else if (l.length > 0 && r.length === 0) return 1;
        else {
            let o = comparePackets(l[0], r[0]);
            if (o !== 0) return o;
            else return comparePackets(l.slice(1), r.slice(1));
        }
    } else if (Array.isArray(l)) {
        return comparePackets(l, [r]);
    } else if (Array.isArray(r)) {
        return comparePackets([l], r);
    } else {
        return l - r;
    }
}

File: AdventOfCode/2024/day13.py
from libs import *

# Parse input

machine_specs = [m.split("\n") for m in read("example").split("\n\n")]
machine_specs = [
    (
        tuple(int(d) for d in a.split("Button A: X+")[1].split(", Y+")),
        tuple(int(d) for d in b.split("Button B: X+")[1].split(", Y+")),
        tuple(int(d) for d in p.split("Prize: X=")[1].split(", Y=")),
    )
    for (a, b, p) in machine_specs
]


# Part 1


def price_to_win(machine, offset=0):
    ((ax, ay), (bx, by), (px, py)) = machine
    px += offset
    py += offset
    # a * ax + b * bx = px => a = (px - b * bx) / ax
    # a * ay + b * by = py => ay * px - b * ay * bx + b * ax * by = py * ax
    b = int((py * ax - ay * px) / (ax * by - ay * bx))
    a = int((px - b * bx) / ax)
    if a * ax + b * bx == px and a * ay + b * by == py:
        return 3 * a + b
    else:
        return 0


part_one(sum([price_to_win(m) for m in machine_specs]))

# Part 2

part_two(sum([price_to_win(m, 10000000000000) for m in machine_specs]))
